llvm/lib/Target/SystemZ/SystemZISelLowering.cpp

```diff
@@ -8826,15 +9029,49 @@ SDValue SystemZTargetLowering::combineSELECT_CCMASK(
   int CCMaskVal = CCMask->getZExtValue();
   SDValue CCReg = N->getOperand(4);
 
-  if (combineCCMask(CCReg, CCValidVal, CCMaskVal))
-    return DAG.getNode(SystemZISD::SELECT_CCMASK, SDLoc(N), N->getValueType(0),
-                       N->getOperand(0), N->getOperand(1),
-                       DAG.getTargetConstant(CCValidVal, SDLoc(N), MVT::i32),
-                       DAG.getTargetConstant(CCMaskVal, SDLoc(N), MVT::i32),
-                       CCReg);
-  return SDValue();
-}
+  if (!combineCCMask(CCReg, CCValidVal, CCMaskVal, DAG))
+    return SDValue();
+
+  // Handle TrueVal and FalseVal in outermost select_ccmask.
+  SDValue TrueVal = N->getOperand(0);
+  SDValue FalseVal = N->getOperand(1);
+  const auto &&TrueSDVals = simplifyAssumingCCVal(TrueVal, CCReg, DAG);
+  const auto &&FalseSDVals = simplifyAssumingCCVal(FalseVal, CCReg, DAG);
+  // There might be cases where TrueSDVals and FalseSDVals are empty as
+  // TrueVal and FalseVal both are non-constant, and they have already been
+  // optimized by combineCCMask, we can not take early exit here, just bypass it
+  // and directly create a new SELECT_CCMASK.
+  if (!TrueSDVals.empty() && !FalseSDVals.empty()) {
+    SmallVector<SDValue, 4> MergedSDVals;
+    for (auto CC : {0, 1, 2, 3}) {
+      MergedSDVals.emplace_back(((CCMaskVal & (1 << (3 - CC))) != 0)
+                                    ? TrueSDVals[CC]
+                                    : FalseSDVals[CC]);
+    }
+    SmallSet<SDValue, 4> MergedSDValsSet(MergedSDVals.begin(),
+                                         MergedSDVals.end());
+    if (MergedSDValsSet.size() != 2)
+      return SDValue();
```

Comment:
I think there are still a few missed optimization opportunities here:
- We can ignore values in the `MergedSDValsSet` for CC values outside of `CCValid`.  This might bring the number of distinct values down to two, when it was originally more.
- Even if we have more than two values, we cannot do *this* optimization, but we still might have done a `combineCCMask` optimization - which we now just ignore.  We shouldn't do that.
- If the `MergedSDValsSet` actually just has a single element, we can completely remove the `SELECT_CCMASK` and simply return that single element here.